### DISCUSSION TITLE
Intercept network activity (request / response) to run assertions

### DISCUSSION THREAD

**Discussion by holek10** [wix/Detox#3122] [Ideas]
An idea to add feature for intercepting network activity - responses & requests - to enable tests that rely on these items.
Similar functionality that already exists in other automation tools out there, like Cypress for example.

**Comment by asafkorem** [comment]
Hey @holek10, thanks for suggesting this idea. I can see why this can be useful.
However, currently, that's something that can be achieved using [mocking](https://wix.github.io/Detox/docs/guide/mocking/).

**Reply by jar0s** [reply]
Hi @asafkorem ,
I believe that @holek10 asks about something a little bit different than mocking. (we work together and talked about it before)
The example he provided is unfortunate as what Cypress calls "intercepting" is in fact mocking.
What he really wants is an option to intercept a network request and/or response in order to run some assertions on them during the test as described [in this blog post](https://www.cypress.io/blog/2019/12/23/asserting-network-calls-from-cypress-tests/#assert-request-and-response-objects).

AFAIK mocking can be used to change the behaviour of the app or mock external systems for the test, but it doesn't provide the ability to verify whether the application sent a HTTP request with a specific payload that we expect. Please correct me if I'm wrong.

The use-case is that the application we work with sends out analytics events when a user interacts with it and we're looking for a way to automatically test/verify the payloads of these analytics events as we had many bugs related to it.

**Reply by asafkorem** [reply]
Hi @jar0s, thanks for the clarification!

I understand your use-case, and the need for such tests, but I'm not sure if such interface should be part of Detox, mainly because it doesn't test user-facing behaviors. Although, I'm not strongly opinionated about that..

Have you considered testing these flows with integration tests / unit tests? I'd also say that it makes sense to use some mocked analytics events sender to track the reported events.

**Reply by jar0s** [reply]
Thank you for your answer @asafkorem .
I did think about using a mocked sender to record the outgoing requests and expose an interface to retrieve them for the purpose of verification/assertions, but before we invested time in building a custom solution, we thought we'd ask in case this was something on your roadmap :wink: 

We'll also explore the option to use integration (or component) tests for this purpose. Could you recommend some good libraries/frameworks in this space? I only know about [RN Testing Lib](https://callstack.github.io/react-native-testing-library/) (which could be useful here), and was wondering if there might be more.

**Reply by asafkorem** [reply]
>I did think about using a mocked sender to record the outgoing requests and expose an interface to retrieve them for the purpose of verification/assertions, but before we invested time in building a custom solution, we thought we'd ask in case this was something on your roadmap 😉

Unfortunately that's not something on our plans, sorry. 😅 
If you'll decide to do so (implement a mocked sender), you can follow this doc: [mocking](https://wix.github.io/Detox/docs/guide/mocking/).

>We'll also explore the option to use integration (or component) tests for this purpose. Could you recommend some good libraries/frameworks in this space? I only know about RN Testing Lib (which could be useful here), and was wondering if there might be more.

It depends, if you kept your views (RN component) separated from the business logic and the app's state (using Redux for instance), you are able to test the app’s logic independently of your RN components, specifically the analytics events reports. If not, I'm not familiar with libraries other than [RNTL](https://callstack.github.io/react-native-testing-library/) for component testing. 

In both cases, [nock](https://github.com/nock/nock) looks promising for HTTP server mocking and expectations, but I haven't examined the other alternatives (there seem to be quite a few more).

**Reply by jar0s** [reply]
Thank you @asafkorem. I really appreciate your input.
I'll take that to the team and we'll look at what might be the best way forward for us.

**Reply by asafkorem** [reply]
Sure thing @jar0s, and thank you both (+@holek10) for the feature idea and the information!
I hope the information I have provided is somehow useful.
And of course, we're here if you need any further information / assistance with Detox 🙂

**Reply by jar0s** [reply]
Hi @asafkorem 
I'd like to follow up on the below topic.

> If you'll decide to do so (implement a mocked sender), you can follow this doc: [mocking](https://wix.github.io/Detox/docs/guide/mocking/).

My understanding, after reading the linked docs is that the mocked sender would be implemented as a mock on the application side. I have an idea how it may look like in general, but I'm not sure how would I go about exposing any data/methods from that mock sender to the process which runs the actual test (outside of the application) so that I could have a meaningful assertion within the test code.

Would you be able to elaborate a bit more on this?

**Reply by asafkorem** [reply]
Hi @jar0s, interesting question actually, I didn't really think of that part.. 

There's no elegant way to expose this data (like in component testing or any "white box" environment), so I'm not familiar with any implementations of such feature for Detox.

You can, for example, use the [getAttributes()](https://wix.github.io/Detox/docs/api/actions-on-element/#getattributes) method to expose the value of a component (`getAttributes().value`), say a transparent modal, that is available only when running with the mocked sender, and set the `value` attribute to some parsable representation of the reported events.

**Reply by asafkorem** [reply]
Another option that came to my mind is to use the mocked sender to post the events over a websocket and to observe them from the tests.

**Reply by jar0s** [reply]
Hi @asafkorem 
Thank you for the suggestions. The one with `getAttributes()` sounds like it could work.
I also like the idea with WebSocket as it may result in a cleaner implementation.
We'll play with them both a bit and see.
Cheers.

**Reply by jar0s** [reply]
Hi @asafkorem ,
I just wanted to let you know (and for anyone else reading this thread in the future) that the solution with WebSocket worked for us.
We start a WS server in the Detox test process and connect to it from a mocked analytics module within the app. The mocked module sends all analytics events through the WebSocket so they can be verified within e2e tests. 

The part with having the WebSocket server running in the same Node process as Detox tests seems pretty generic so I wonder if this might be something worth contributing back to the Detox itself to enable communication between the app and the test process in general? What do you think?
How we've done it is that the tests can subscribe to the topics of interest and WS server broadcasts the messages to specific subscribers based on the topics (and this mechanism is pretty generic as well).

**Reply by asafkorem** [reply]
Hi @jar0s, sorry for noticing this in such delay. Glad to hear that it worked out for you!
I'm not sure if such feature should be part of Detox itself. However, we already have some contribution that does that https://github.com/wix/Detox/pull/3245, and based on the demand, I guess it worth thinking about this.

**Reply by jar0s** [reply]
Hi @asafkorem 
I think #3245 solves a "reverse" challenge. It allows injecting something into the app (detox test process -> the app). What we needed and implemented with WebSocket is the app sending info to the Detox test process (the app -> detox test process).
I see both as supplementary solutions serving different use cases.

**Reply by asafkorem** [reply]
Thanks for letting us know that the issue has been resolved, @jar0s. If this is still relevant for you or anyone else reading this thread, you may want to consider using [MSW](https://github.com/mswjs/msw) as a solution. It seems that someone has successfully used it with Detox tests, as mentioned in this [GitHub issue](https://github.com/wix/Detox/issues/3808).
